Update `bdk_electrum` crate to use sync/full-scan structs

LagginTimes opened the pull request:
Fixes #1265
Possibly fixes #1419

### Context

Previous changes such as

* Universal structures for full-scan/sync (PR #1413)
* Making `CheckPoint` linked list query-able (PR #1369)
* Making `Transaction`s cheaply-clonable (PR #1373) 

has allowed us to simplify the interaction between chain-source and receiving-structures (`bdk_chain`).

The motivation is to accomplish something like this ([as mentioned here](https://github.com/bitcoindevkit/bdk/issues/1153#issuecomment-1752263555)):
```rust
let things_I_am_interested_in = wallet.lock().unwrap().start_sync();
let update = electrum_or_esplora.sync(things_i_am_interested_in)?;
wallet.lock().unwrap().apply_update(update)?:
``` 

### Description

This PR greatly simplifies the API of our Electrum chain-source (`bdk_electrum`) by making use of the aforementioned changes. Instead of referring back to the receiving `TxGraph` mid-sync/scan to determine which full transaction to fetch, we provide the Electrum chain-source already-fetched full transactions to start sync/scan (this is cheap, as transactions are wrapped in `Arc`s since #1373).

In addition, an option has been added to include the previous `TxOut` for transactions received from an external wallet for fee calculation.

### Changelog notice

* Change `TxGraph::insert_tx` to take in anything that satisfies `Into<Arc<Transaction>>`. This allows us to reuse the `Arc` pointer of what is being inserted.
* Add `tx_cache` field to `SyncRequest` and `FullScanRequest`.
* Make `Anchor` type in `FullScanResult` generic for more flexibility.
* Change `ElectrumExt` methods to take in `SyncRequest`/`FullScanRequest` and return `SyncResult`/`FullScanResult`. Also update electrum examples accordingly.
* Add `ElectrumResultExt` trait which allows us to convert the update `TxGraph` of `SyncResult`/`FullScanResult` for `bdk_electrum`.
* Added an option for `full_scan` and `sync` to also fetch previous `TxOut`s to allow for fee calculation.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

LagginTimes committed:
refactor(electrum): remove `RelevantTxids` and track txs in `TxGraph`

This PR removes `RelevantTxids` from the electrum crate and tracks
transactions in a `TxGraph`. This removes the need to separately
construct a `TxGraph` after a `full_scan` or `sync`.

ValuedMammal commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -373,10 +264,16 @@ fn construct_update_tip(
+) -> Option<ConfirmationTimeHeightAnchor> {
+    let confirmation_time = client
+        .block_header(raw_height as usize)
+        .expect("header must exist")
```
🔧 I think we should handle the error the way we do for the other `ElectrumApi` calls, but since `determine_tx_anchor` doesn't return a `Result`, we could try doing the api call outside the function and pass in the header time on its own, or maybe as a (height, time) tuple.

LagginTimes replied:
I've been kind of playing around with using a batch command but it's kind of a mess right now, but I'll push a simple fix with your suggestion as it is more clean.

ValuedMammal reviewed (commented):
Looking good so far.

I made a note related to adding the `client` as a parameter to `determine_tx_anchor` - you may have already considered this. I see you mentioned potentially doing more batch calls, which is a good idea assuming we can ensure the results are in order consistent with the request.

side note: the commit message could use an extra newline after the subject and before the body.

evanlinjin reviewed (commented):
~~It's great that the API is simplified now. However, with the current changes, every call to `ElectrumExt::full_scan` forces us to download every full tx?~~

**EDIT:** Ahh, I see your TODOs in your description.

evanlinjin commented:
For the commit message, remember to put a new line between first line and body.

evanlinjin commented:
It seems that you intended the sync flow to be:
1. Copy the wallet's `TxGraph` and pass it to the chain source.
2. Perform update, and populate `TxGraph` with missing transactions/anchors and return it back to the wallet.

However, the downsides of this approach is the following:
1. Because we are cloning the wallet's `TxGraph`, and `TxGraph` stores both transactions and anchors, and only transactions are cheaply copied (not anchors), therfore copying the `TxGraph` over is non-optimal.
2. The anchors that are copied from the wallet cannot be used anyway. We need to keep the anchor type in the `TxGraph` a generic to support all wallet types. However, we wish to return an update `TxGraph` with the `ConfirmationTimeHeightAnchor` to represent all information available from Electrum. We cannot guarantee that the wallet's anchor type can be transformed into `ConfrimastionTimeHeightAnchor`, therefore we end up just re-fetching the anchors from Electrum anyway.

Hence, I think it is better to have a transaction cache `HashMap<Txid, Arc<Transaction>>` and the `TxGraph` update as a separate concept.

evanlinjin committed:
feat(chain): `TxGraph::insert_tx` reuses `Arc`

When we insert a transaction that is already wrapped in `Arc`, we should
reuse the `Arc`.

evanlinjin committed:
feat(electrum)!: use new sync/full-scan structs for `ElectrumExt`

`ElectrumResultExt` trait is also introduced that adds methods which can
convert the `Anchor` type for the update `TxGraph`.

We also make use of the new `TxCache` fields in
`SyncRequest`/`FullScanRequest`. This way, we can avoid re-fetching full
transactions from Electrum if not needed.

Examples and tests are updated to use the new `ElectrumExt` API.

evanlinjin committed:
feat(chain): introduce `TxCache` to `SyncRequest` and `FullScanRequest`

This transaction cache can be provided so the chain-source can avoid
re-fetching transactions.

evanlinjin committed:
feat(wallet): cache txs when constructing full-scan/sync requests

evanlinjin committed:
fix(chain): Make `Anchor` type in `FullScanResult` generic

evanlinjin committed:
feat: update logging of electrum examples

* Syncing with `example_electrum` now shows progress as a percentage.
* Flush stdout more aggressively.

LagginTimes commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -1,159 +1,44 @@
     ///
     /// - `prev_tip`: the most recent blockchain tip present locally
     /// - `keychain_spks`: keychains that we want to scan transactions for
+    /// - `full_txs`: [`TxGraph`] that contains all previously known transactions
```
Docs here will need to be updated.

LagginTimes commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -1,159 +1,44 @@
     /// - `prev_tip`: the most recent blockchain tip present locally
     /// - `misc_spks`: an iterator of scripts we want to sync transactions for
-    /// - `txids`: transactions for which we want updated [`Anchor`]s
+    /// - `full_txs`: [`TxGraph`] that contains all previously known transactions
```
Here as well.

LagginTimes commented on crates/electrum/src/lib.rs:
```diff
@@ -7,19 +7,10 @@
-//! 1. Determine which full transactions are missing. Use [`RelevantTxids::missing_full_txs`].
-//!
-//! 2. Obtaining the full transactions. To do this via electrum use [`ElectrumApi::batch_transaction_get`].
+//! [`bdk_chain`] including [`bdk_chain::TxGraph`], which includes `txid`s and full transactions.
```
Also here.

LagginTimes committed:
docs(electrum): fixed `full_scan`, `sync`, and crate documentation

LagginTimes committed:
fix(electrum): fixed `sync` functionality

evanlinjin commented:
@LagginTimes thanks for cleaning it up!

ValuedMammal commented on crates/chain/src/spk_client.rs:
```diff
@@ -204,10 +240,35 @@ impl<K: Ord + Clone> FullScanRequest<K> {
     }
 
+    /// Add to the [`Transaction`] cache that allows the chain source to avoid re-fetching full
+    /// transactions.
```
I think it's enough to state the purpose of `TxCache` (to avoid re-fetching) once in the documentation for the type itself, rather than every time it makes an appearance, lol.

```suggestion
    /// Add to the [`Transaction`] cache held by the request.
```

evanlinjin replied:
Fixed.

ValuedMammal commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -245,54 +127,129 @@ impl<A: ElectrumApi> ElectrumExt for A {
+///
+/// Currently, only a single method exists that converts the update [`TxGraph`] to have an anchor
+/// type of [`ConfirmationTimeHeightAnchor`].
+pub trait ElectrumResultExt {
```
While I see its utility, I don't love this from an API perspective. It feels like an after-thought instead of something that should be handled within a `sync` or `full_scan`.

📌 I'd like to find a way for the `ElectrumExt` methods to return a TxGraph with confirmation time anchors by default, then the caller can always `map_anchors` on the update graph to a simpler anchor type without needing the client.

But right now I don't see an easy solution other than to 1) try converting into time-anchors during `full_scan`, 2) reduce them back to height-anchors in `sync` before populating with txids and outpoints, and 3) convert them again before returning the `SyncResult`.

evanlinjin replied:
@LagginTimes do you have any ideas on this one?

evanlinjin replied:
The reason why `sync`/`full_scan` returns updates with `ConfirmationHeightAnchor` is because this is what the Electrum API provides us with. To get the block time requires extra calls, hence, I think it makes sense to put this in a separate method. I agree that using the `ElectrumResultExt` trait is clunky. Maybe a better option will be to first return a custom type? And methods on the custom type returns `SyncResult`/`FullScanResult` per the caller's requirement.

evanlinjin replied:
@ValuedMammal I think you will be happy with this solution: https://github.com/bitcoindevkit/bdk/pull/1403/commits/7a69d359135efc2898c318e90479dd03791962ea

ValuedMammal replied:
Maybe I'm bikeshedding, but what if we just made this another `ElectrumExt` method that does the same thing as the current `try_into_confirmation_time_result` and merely converts the first graph update to a new TxGraph containing confirmation time anchors, with no need to introduce a new type.

```rust
    /// Converts the given [`TxGraph`] into one with anchors of type [`ConfirmationTimeHeightAnchor`].
    fn confirmation_time_tx_graph(
        &self,
        graph: TxGraph<ConfirmationHeightAnchor>,
    ) -> Result<TxGraph<ConfirmationTimeHeightAnchor>, Error>;
```

ValuedMammal commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -497,18 +455,36 @@ fn populate_with_txids(
-        let tx_entry = relevant_txids.0.entry(txid).or_default();
+        if graph_update.get_tx(txid).is_none() {
+            // TODO: We need to be able to insert an `Arc` of a transaction.
+            let _ = graph_update.insert_tx(tx);
```
This TODO is now done, right?

ValuedMammal commented on example-crates/example_electrum/src/main.rs:
```diff
@@ -147,42 +148,47 @@ fn main() -> anyhow::Result<()> {
+                        let mut once = BTreeSet::new();
+                        move |k, spk_i, _| {
+                            if once.insert(k) {
+                                eprint!("\nScanning {}: {}", k, spk_i);
```
```suggestion
                                eprint!("\nScanning {}: {} ", k, spk_i);
```

Without this, spk index 0 and 1 are stuck together. Or you could make it a leading space on L177.

ValuedMammal commented on example-crates/wallet_electrum/src/main.rs:
```diff
@@ -38,44 +39,27 @@ fn main() -> Result<(), anyhow::Error> {
+            let mut once = HashSet::<KeychainKind>::new();
+            move |k, spk_i, _| match once.insert(k) {
+                true => print!("\nScanning keychain [{:?}]", k),
+                false => print!(" {:<3}", spk_i),
```
nit: matching on a bool when an if/else would suffice.

ValuedMammal reviewed (commented):
Overall I'm a fan of `TxCache` and I think this is a solid improvement. These are my lingering thoughts:

- What do you see as the next logical step to improving the electrum crate, for example in relation to #1332 and #1265 
- In electrum_ext: do we really need the hard-coded genesis txid? Is this condition even possible?
- In electrum_ext: Can we add comments for the `populate_with_*` functions? what's being populated, who's doing the populating, etc

evanlinjin commented:
> What do you see as the next logical step to improving the electrum crate, for example in relation to [Rearchitect `bdk_electrum` syncing/scanning. #1332](https://github.com/bitcoindevkit/bdk/issues/1332) and [Option for Electrum and Esplora APIs to include floating `TxOut`s in updates for fee calculation #1265](https://github.com/bitcoindevkit/bdk/issues/1265)

I would tacke #1265 first. This will make `bdk_electrum` "feature complete". We can worry about the internals of it after 1.0 is released.

> In electrum_ext: do we really need the hard-coded genesis txid? Is this condition even possible?

I would assume not, however someone needs to have a look into it. Additionally, the hardcoded txid is not complete as it only includes the genesis coinbase tx for mainnet. 

> In electrum_ext: Can we add comments for the `populate_with_*` functions? what's being populated, who's doing the populating, etc

Good idea.

LagginTimes commented:
I will start poking around with the electrum side of #1265 as well.

evanlinjin committed:
feat(chain)!: use custom return types for `ElectrumExt` methods

This is more code, but a much more elegant solution than having
`ElectrumExt` methods return `SyncResult`/`FullScanResult` and having an
`ElectrumResultExt` extention trait.

LagginTimes committed:
feat(electrum): include option for previous `TxOut`s for fee calculation

The previous `TxOut` for transactions received from an external
wallet may be optionally added as floating `TxOut`s to `TxGraph`
to allow for fee calculation.

evanlinjin commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -484,6 +507,26 @@ fn fetch_tx<C: ElectrumApi>(
+    }
+    Ok(())
+}
+
```
@LagginTimes This logic is wrong. We only want to calculate fees for relevant transactions, not every single transaction stored in cache. The cache may be shared between multiple chain-sources and wallets.

evanlinjin replied:
Also a comment will be nice (I should also add comments to all the other methods).

LagginTimes replied:
Thank you for the clarification, hopefully everything looks better now.

ValuedMammal replied:
I like how you had this originally, where the function takes a `&mut TxGraph` and returns nothing (cloning only the `TxOut` to be inserted).

```rust
fn fetch_prev_txout<C: ElectrumApi>(
    client: &C,
    tx_cache: &mut TxCache,
    graph: &mut TxGraph<ConfirmationHeightAnchor>,
) -> Result<(), Error> {
    let full_txs: Vec<Arc<Transaction>> = graph.full_txs().map(|tx_node| tx_node.tx).collect();
    for tx in full_txs {
        for vin in &tx.input {
            let outpoint = vin.previous_output;
            let prev_tx = fetch_tx(client, tx_cache, outpoint.txid)?;
            for txout in prev_tx.output.clone() {
                let _ = graph.insert_txout(outpoint, txout);
            }
        }
    }
    Ok(())
}
```

evanlinjin replied:
This is a good solution, thanks @ValuedMammal

evanlinjin commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -38,13 +40,19 @@ pub trait ElectrumExt {
     /// `batch_size` specifies the max number of script pubkeys to request for in a single batch
-    /// request.
+    /// request. `fetch_prev_txouts` specifies whether or not we want previous `TxOut`s for fee
+    /// calculation.
```
Let's have some consistency here. I think we should either have all input descriptions as bullet-points or no bullet-points. Same for `full_scan` docs.

LagginTimes replied:
I initially tried to keep the original formatting as I assumed maybe there was a reason for it, but it sounds like there wasn't. 🤔

evanlinjin commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -104,6 +113,18 @@ impl<E: ElectrumApi> ElectrumExt for E {
+                        .collect::<Vec<_>>(),
+                )?);
+            }
+
```
Let's do this AFTER the reorg check. This is safe as we are fetching full transactions (monotonic data). Doing this AFTER the reorg check can, potentially save bandwidth (when a reorg happens).

evanlinjin replied:
Nit: Styling wise, I would create a new line for the call to `fetch_prev_txout`.

evanlinjin replied:
Or we can do what @ValuedMammal suggested in https://github.com/bitcoindevkit/bdk/pull/1403#discussion_r1593887715 (which I also think is better).

evanlinjin commented on crates/electrum/src/electrum_ext.rs:
```diff
@@ -165,6 +191,19 @@ impl<E: ElectrumApi> ElectrumExt for E {
+                    .map(|tx_node| tx_node.tx)
+                    .collect::<Vec<_>>(),
+            )?);
+        }
```
We don't want to `fetch_prev_txout` twice (the first time is done in `full_scan` which we call here). Fetching twice is wasteful as we don't need to calculate fee of parents of relevant transactions.

ValuedMammal replied:
I had the same thought, but then wondered if there might be new relevant data as a result of populating with txids and outpoints?

evanlinjin replied:
Yes, bool should be false initially for the full_scan call and we fetch prev txouts at the end.

evanlinjin reviewed (approved):
ACK 4ccd86aaf67c3d0e4d5621898be138941aab84cd

evanlinjin commented:
A test to see whether fee is actually calculated will be even better.

evanlinjin committed:
chore(electrum): do not use `anyhow::Result` directly

evanlinjin commented:
This needs a rebase, merging #1203 broke this.

evanlinjin committed:
feat(electrum): update docs and simplify logic of `ElectrumExt`

Helper method docs are updated to explain what they are updating. Logic
is simplified as we do not need to check whether a tx exists already in
`update_graph` before inserting it.

ValuedMammal reviewed (approved):
ACK b45897e6fe2f7e67f5d75ec6f983757b28c5ec19

notmandatory reviewed (approved):
ACK b45897e6fe2f7e67f5d75ec6f983757b28c5ec19

Tested with examples: `example_electrum` and `wallet_electrum`
